### PR TITLE
CS.R.50 Switch statement

### PR DESCRIPTION
**Pull request by finomen**
Each case statement MUST be a scoped statement.

Relates: T10105

### CODE REVIEW

File: clang-tools-extra/clang-tidy/daedalean/SwitchStatementCheck.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+namespace clang {
+namespace tidy {
+namespace daedalean {
+
+void SwitchStatementCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(switchStmt().bind("x"), this);
+}
+
+void SwitchStatementCheck::check(const MatchFinder::MatchResult &Result) {
+  // FIXME: Add callback implementation.
```
Comment (aburgm) [inline clang-tools-extra/clang-tidy/daedalean/SwitchStatementCheck.cpp]: what does this mean?

File: clang-tools-extra/test/clang-tidy/checkers/daedalean-switch-statement.cpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  switch (a) {
+  case 1:
+    // CHECK-MESSAGES: :[[@LINE-1]]:9: warning: Case statement must be scoped statement [daedalean-switch-statement]
+  case 2:
+    // CHECK-MESSAGES: :[[@LINE-1]]:9: warning: Case statement must be scoped statement [daedalean-switch-statement]
+    bar();
+    break;
+  case 3: {
+    bar();
+  }
```
Comment (aburgm) [inline clang-tools-extra/test/clang-tidy/checkers/daedalean-switch-statement.cpp:16]: what about explicit fallthrough? we don't want something like this to be flagged:

```
switch(x) {
case 1:
  [[ fallthrough ]]
case 2: {
    bla();
  }
}
```

Maybe not even without the fallthrough, and then have another check for missing break or fallthrough at the end of a case

File: clang-tools-extra/docs/clang-tidy/checks/daedalean-switch-statement.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+.. title:: clang-tidy - daedalean-switch-statement
+
+daedalean-switch-statement
+==========================
+
+FIXME: Describe what patterns does the check detect and why. Give examples.
```
Comment (aburgm) [inline clang-tools-extra/docs/clang-tidy/checks/daedalean-switch-statement.rst]: here also a short explanation would be better than this fixme

File: clang-tools-extra/test/clang-tidy/checkers/daedalean-switch-statement.cpp (hunk 2 above)
Comment (finomen) [reply clang-tools-extra/test/clang-tidy/checkers/daedalean-switch-statement.cpp:16]: -Wimplicit-fallthrough compiler flag checks exactly this

Comment (aburgm) [reply clang-tools-extra/test/clang-tidy/checkers/daedalean-switch-statement.cpp:16]: As discussed: fallthrough is a statement, so we handle it as such and require it to be scoped as well.

**Review by aburgm — DISMISSED** [review]

**Review by aburgm — DISMISSED** [review]

**Review by aburgm — APPROVED** [review]

**Pull request merged**